src/sql/workbench/parts/objectExplorer/browser/serverTreeRenderer.ts

```diff
@@ -152,19 +158,72 @@ export class ServerTreeRenderer implements IRenderer {
 		let iconLowerCaseName = iconName.toLocaleLowerCase();
 		templateData.icon.classList.add(iconLowerCaseName);
 
+		if (treeNode.iconPath) {
+			iconRenderer.putIcon(templateData.icon, treeNode.iconPath);
+		}
+
 		templateData.label.textContent = treeNode.label;
 		templateData.root.title = treeNode.label;
 	}
 
+	private getContextKeyService(connection: ConnectionProfile): IContextKeyService {
+		if (!connection || !connection.id) { return undefined; }
+		let serverInfo: ServerInfo = this._connectionManagementService.getServerInfo(connection.id);
+		if (!serverInfo) { return undefined; }
+
+		let scopedContextService: IContextKeyService = this._contextKeyService.createScoped();
+		let serverInfoContextKey = new ServerInfoContextKey(scopedContextService);
+		serverInfoContextKey.set(serverInfo);
+		return scopedContextService;
+	}
+
+	private getIconPath(connection: ConnectionProfile): IconPath {
+		let providerProperties = this._connectionManagementService.getProviderProperties(connection.providerName);
+		if (!providerProperties) { return undefined; }
+
+		let iconPath: IconPath = undefined;
+		let pathConfig: URI | IconPath | { when: string, path: IconPath }[] = providerProperties['iconPath'];
+		if (Array.isArray(pathConfig)) {
+			let scopedContextService: IContextKeyService = this.getContextKeyService(connection);
+			if (!scopedContextService) { return undefined; }
+			for (const e of pathConfig) {
+				let when = ContextKeyExpr.deserialize(e.when);
```

Comment:
When should be optional - if no "when" is defined, shouldn't we just use the icon?

Reply:
fixed.